C/dag.c

```diff
@@ -136,6 +136,61 @@ static sha256_midstate wmrIV(int32_t tag) {
   return (sha256_midstate){0};
 }
 
+/* Given a SHA-256 midstate, 's', of '*count / 512' blocks, and
+ * a 'block' with '*count % 512' bits set and with the remaining bits set to 0,
+ * push a new bit of value 'bit' onto the block.
+ * If the block ends up full, compress and update the midstate.
+ *
+ * Precondition: uint32_t s[8];
+ *               uint32_t block[16];
+ *               NULL != count;
+ */
+static void sha256_pushBit(uint32_t* s, uint32_t* block, size_t* count, bool bit) {
+  if (bit) block[*count / 32 % 16] |= (uint32_t)1 << (31 - *count % 32);
+  (*count)++;
+  if (0 == *count % 512) {
+    sha256_compression(s, block);
+    memset(block, 0, sizeof(uint32_t[16]));
+  }
+}
+
+/* Given a SHA-256 midstate, 's', of '*count / 512' blocks, and
+ * a 'block' with '*count % 512' bits set and with the remaining bits set to 0,
+ * finalize the SHA-256 computation by adding SHA-256 padding and set 's' to the resulting SHA-256 hash.
+ *
+ * Precondition: uint32_t s[8];
+ *               uint32_t block[16];
+ *               NULL != count;
+ */
+static void sha256_end(uint32_t* s, uint32_t* block, const size_t len) {
+  size_t count = len;
+  sha256_pushBit(s, block, &count, 1);
+  if (448 < count % 512) {
```

Comment:
To be clear - the use of `<` here rather than `<=` is simply wrong, because in the `==` case the bit you just pushed will be overwritten by the length?

(I notice you fixed this in a later commit.)

Reply:
I believe the use of `<` is correct here.  If the count is exactly 448 then message length is 447 bits, and only a [message that is 448 bits or larger is padded beyond the block size](https://crypto.stackexchange.com/a/29851/47684) requires running another iteration of the compression function.

The bit is not overwritten because after writing the 448th bit, there are still 64 bits in the 512 bit block, which is enough space to write the length suffix without overwriting the bit.

Reply:
However, if this were an error, I do want to fix it, even if it is only part of history.  I have a specific commit that is intended as only an optimization, not a change in functionality.

Reply:
Ah! Yep, I missed that `count` is one more than the message length.